### ISSUE TITLE
Object access reports error after upgrading to TypeScript 5.5

### ISSUE THREAD

**Issue by jkillian** [microsoft/TypeScript#59185]
### 🔎 Search Terms

"noUncheckedIndexedAccess", "Object is possibly 'undefined'."

### 🕗 Version & Regression Information

- This changed between versions 5.4.5 and 5.5.3

### ⏯ Playground Link

https://www.typescriptlang.org/play/?noUncheckedIndexedAccess=true&ts=5.5.3#code/DYUwLgBAzmBOBc04EsB2BzCAfCBXVAJiAGZogEDcAUMsRABQCyAhmABYB0szhA9gLb0AlBAA8EAAwcAjCIDeVCElgQAvBABEzDVQC+VKgGNeqGBGaHDiORADaqZvxCIYsNOgC6iVLn4AjEBVdNQg5XWpaBlcIAEJ1H2BgeUVzS1tXDwgAfiz1CWolC0N0uA8AajLqfSA

### 💻 Code

```ts
let str: string | undefined;
if (Math.random() < 0.1) {
  str = "a"
}

const acc: { [name: string]: number } = {};
if (str != null) {
  acc[str] ??= 0;
  acc[str]++; // In TS 5.5: Object is possibly 'undefined'. (2532)
}
```

### 🙁 Actual behavior

In TS 5.5.x, with `noUncheckedIndexedAccess` enabled, `acc[str]++;` reports an error "Object is possibly 'undefined'. (2532)".

### 🙂 Expected behavior

In TS 5.4.x and before, no errors were reported.

### Additional information about the issue

I admittedly didn't do a very deep dive on the root cause here, but wanted to file an issue because intuitively it felt unexpected when upgrading to TS 5.5 that the above code now fails type-checking. Apologies if I missed an existing issue for this or something in the release notes!

**Comment by andrewbranch** [comment]
I highly doubt that, @typescript-bot

**Comment by Andarist** [comment]
5.4 incorrectly always allowed `acc[str]++`:
```ts
let str: string | undefined;
if (Math.random() < 0.1) {
  str = "a"
}

const acc: { [name: string]: number } = {};
if (str != null) {
  // acc[str] ??= 0;
  acc[str]++; // no error in 5.4
}
```

This has been fixed in 5.5 and that affects ur repro. TS 5.5 introduces [Control Flow Narrowing for Constant Indexed Accesses](https://devblogs.microsoft.com/typescript/announcing-typescript-5-5/#control-flow-narrowing-for-constant-indexed-accesses) but we can read there that:
> TypeScript is now able to narrow expressions of the form obj[key] when both obj and key are effectively constant.

Since your `str` isn't effectively constant... this doesn't work in a form that u'd like it to. Notice how this version works alright:
```ts
let str = Math.random() < 0.1 ? "a" : undefined;

const acc: { [name: string]: number } = {};

if (str != null) {
  acc[str] ??= 0;
  acc[str]++;
}

export {};
```

**Comment by jkillian** [comment]
Thanks for some of the context @Andarist! Perhaps that makes this a feature request instead of a bug report then. It does seem like for the duration of these three lines, `str` is effectively constant enough that TS could allow the code regardless of the assignment above:

```ts
if (str != null) {
  acc[str] ??= 0;
  acc[str]++;
}
```

Does that seem right, or am I missing something here?

**Comment by Andarist** [comment]
From what I understand, it was a performance concern to allow further analysis to handle cases like those. But yes - it is a potential feature request.
